src/libs/viewpoints/viewpoint_extractor.ts

```diff
@@ -0,0 +1,536 @@
+
+    /**Step 1 Get all Roles in model
+     * 
+     * Step 2 Get all sortal ancestors 
+     * 
+     * Step 3 Get all relations between all classes (roles and sortal ancestors)
+     * 
+     * Step 4 Get all generalizations between all classes (roles and sortal ancestors)
+     */
+
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.ROLE);
+
+    let generals = [];
+    for (let i = 0; i < classes.length; i++) {
+      generals = generals.concat(classes[i].getAncestors().filter((cl) => cl.hasSortalStereotype()));
+    }
+
+    viewpoint.addClasses(classes);
+    viewpoint.addClasses(generals);
+
+    viewpoint.removeDuplicates();
+
+    viewpoint.addRelations(this.getRelationsBetweenClasses(viewpoint.classes));
+    viewpoint.addGeneralizations(this.getGeneralizationsBetweenClasses(viewpoint.classes));
+
+    viewpoint.removeDuplicates();
+
+    return viewpoint;
+  }
+
+  /**
+     * This method filter all Phases, respective sortal ancestors  and related relations and group them in a view
+     * 
+     * @returns A viewpoint extracted using the Phase Pattern
+     */
+  phaseView(): Viewpoint {
+    const viewpoint = new Viewpoint('Phase View');
+
+
+    /**Step 1 Get all Phases in model
+     * 
+     * Step 2 Get all sortal ancestors 
+     * 
+     * Step 3 Get all relations between all classes (phases and sortal ancestors)
+     * 
+     * Step 4 Get all generalizations between all classes (phases and sortal ancestors)
+     */
+    const classes = this.project.getAllClassesByStereotype(ClassStereotype.PHASE);
+
+    let generals = [];
+    for (let i = 0; i < classes.length; i++) {
+      generals = generals.concat(classes[i].getAncestors().filter((cl) => cl.hasSortalStereotype()));
+    }
+
+    viewpoint.addClasses(classes);
+    viewpoint.addClasses(generals);
+
+    viewpoint.removeDuplicates();
+
+    viewpoint.addRelations(this.getRelationsBetweenClasses(viewpoint.classes));
+    viewpoint.addGeneralizations(this.getGeneralizationsBetweenClasses(viewpoint.classes));
+
+    viewpoint.removeDuplicates();
+
+    return viewpoint;
+  }
+
+  /**
+     * This method filter all Relators, respective direct conectect clases by mediator relation, their sortal ancestors  
+     * and related relations and group them in a view
+     * 
+     * @returns A viewpoint extracted using the Relator Pattern
+     */
+  relatorView = () => {
+    const viewpoint = new Viewpoint('Relator View');
+
+
+    /**Step 1 Get all Relators in model
+     * 
+     * Step 2 Get all classes conected by mediation relation 
+     * 
+     * Step 3 Get all sortal ancestors 
+     * 
+     * Step 4 Get all relations between all classes 
+     * 
+     * Step 5 Get all generalizations between all classes 
+     */
+    let classes = this.project.getAllClassesByStereotype(ClassStereotype.RELATOR);
+
+    const relacoes = this.project.getAllRelations();
+    let mediacoes = [];
+
+    for (let j = 0; j < classes.length; j++) {
```

Comment:
Prefer using a for-in loop (https://www.w3schools.com/js/js_loop_forin.asp). Please check other usages.

Reply:
Done